### PR TITLE
Flip

### PR DESCRIPTION
**Pull request by davidsoncolin**
Added a flip_move function and a "flip" uci-like command which swaps sides for the entire game history.
Added test to ensure flip_move works as per the previous code; extended it to work on edge cases.
Added test to ensure the Network::gather_features function is invariant to flipping (it wasn't).
Fixed a bug in Network::gather_features to make it invariant. If the moves are presented to the network flipped as if white's turn to move then it makes no sense to also tell the network whose turn it is to move. Eval should be invariant to flip.

Renamed type_of to move_type

### CODE REVIEW

**Comment by davidsoncolin** [comment]
I'm going to comment and, for once, not accidentally press the "close and comment" button.

The network gives slightly different evaluations on the same data whether it is white or black to move. This is because one bit of data - "whose turn" - is set when it is black's turn to move. When the change was made to always flip the board so that it is presented with the player to move going up, that bit of data became meaningless. Since the promotion bug primarily corrupted black's moves, it seems that fixing this bit might get us back on track faster.

**Comment by killerducky** [comment]
This PR deviates from the Deepmind paper, and changes inputs to the NN. I won't pull this so close to after a major change to the NN. And I won't pull it without consulting @glinscott and @Error323. I'm not sure I think it's a good idea at all, will have to think about this more.

File: src/Types.h [hunk 1]
```diff
@@ -346,7 +346,7 @@ constexpr int from_to(Move m) {
  return m & 0xFFF;
 }
 
-constexpr MoveType type_of(Move m) {
+constexpr MoveType move_type(Move m) {
```
Comment (glinscott) [inline src/Types.h]: Let's move the renaming to another PR please - it just complicates things.

Comment (glinscott) [inline src/Types.h]: Also, this comes from SF - unless there is a reallllly strong reason, I'd rather leave it the same.
